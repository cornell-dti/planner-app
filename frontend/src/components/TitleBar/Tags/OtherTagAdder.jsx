// @flow strict

import React from 'react';
import { connect } from 'react-redux';
import { Icon } from 'semantic-ui-react';
import { GithubPicker } from 'react-color';
import type { Tag } from '../../../store/store-types';
import styles from './TagItem.css';
import type { AddTagAction } from '../../../store/action-types';
import { addTag as addTagAction } from '../../../store/actions';
import { randomId } from '../../../util/general-util';
import { colMap } from './ListColors';

type Props = {|
  +addTag: (tag: Tag) => AddTagAction
|};

type State = {|
  +showEditor: boolean;
  +color: string;
  +reset: boolean;
|};

const colArray: string[] = Object.keys(colMap);

class OtherTagAdder extends React.Component<Props, State> {
  state: State = {
    // whether to show the name editor and color picker
    showEditor: false,
    color: '#56d9c1',
    reset: true,
  };

  toggleEditor = () => {
    this.setState(s => ({ ...s, showEditor: !s.showEditor }));
  };

  handleColor = (e) => {
    this.setState({ color: e.hex, reset: false });
    this.toggleEditor();
  };

  handleSave = (saveEvent: SyntheticKeyboardEvent<HTMLInputElement>) => {
    if (saveEvent.keyCode !== 13) {
      return;
    }
    const currName = saveEvent.currentTarget.value;
    const { color } = this.state;
    const currColor = color;

    this.setState({ color: '#56d9c1', reset: true, showEditor: false });
    saveEvent.currentTarget.value = '';
    const { addTag } = this.props;
    addTag({
      id: randomId(),
      type: 'other',
      name: currName,
      color: currColor,
    });
  };

  render() {
    const { showEditor, reset, color } = this.state;
    return (
      <li className={styles.ColorConfigItem}>
        <input
          type="text"
          className={`${styles.TagName} ${styles.AddTagName}`}
          placeholder="New Tag"
          onKeyDown={this.handleSave}
        />
        <button type="button" className={styles.ColorEdit} onClick={this.toggleEditor}>
          {reset ? 'Select' : colMap[color.toLowerCase()]}
          <span className={styles.ColorEditDisp} style={{ backgroundColor: color }} />
          <Icon className={styles.ColorEditArrow} name="angle down" />
        </button>
        {showEditor && (
          <div className={styles.OpenPicker}>
            <GithubPicker
              color={color}
              onChangeComplete={this.handleColor}
              triangle="top-right"
              colors={colArray}
            />
          </div>
        )}
      </li>
    );
  }
}

const ConnectedOtherTagAdder = connect(null, { addTag: addTagAction })(OtherTagAdder);
export default ConnectedOtherTagAdder;
