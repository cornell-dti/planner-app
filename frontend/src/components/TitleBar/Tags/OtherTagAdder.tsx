import React, { KeyboardEvent, SyntheticEvent, ReactElement } from 'react';
import styles from './TagItem.module.scss';
import ColorEditor from './ColorEditor';
import { addTag } from '../../../firebase/actions';

type State = { readonly name: string; readonly color: string };

const defaultColor = '#289de9';
const initialState: State = { name: '', color: defaultColor };

export default class OtherTagAdder extends React.PureComponent<unknown, State> {
  public readonly state: State = initialState;

  private editColor = (color: string): void => this.setState({ color });

  private editName = (event: SyntheticEvent<HTMLInputElement>): void =>
    this.setState({
      name: event.currentTarget.value,
    });

  private onSubmit = (event: KeyboardEvent<HTMLInputElement>): void => {
    if (event.key !== 'Enter') {
      return;
    }
    const { name, color } = this.state;
    addTag({
      name,
      color,
      classId: null,
    });
    this.setState(initialState);
  };

  public render(): ReactElement {
    const { name, color } = this.state;
    return (
      <li className={styles.ColorConfigItem}>
        <input
          type="text"
          className={`${styles.TagName} ${styles.AddTagName}`}
          placeholder="New Tag"
          value={name}
          onChange={this.editName}
          onKeyDown={this.onSubmit}
        />
        <ColorEditor color={color} onChange={this.editColor} />
      </li>
    );
  }
}
