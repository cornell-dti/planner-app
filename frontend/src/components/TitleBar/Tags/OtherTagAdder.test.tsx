import React from 'react';
import renderer from 'react-test-renderer';
import OtherTagAdder from './OtherTagAdder';

it('OtherTagAdder matches snapshot.', () => {
  const tree = renderer.create(<OtherTagAdder />).toJSON();
  expect(tree).toMatchSnapshot();
});
