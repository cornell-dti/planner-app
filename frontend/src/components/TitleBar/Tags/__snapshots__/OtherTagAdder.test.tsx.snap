// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OtherTagAdder matches snapshot. 1`] = `
<li
  className="ColorConfigItem"
>
  <input
    className="TagName AddTagName"
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="New Tag"
    type="text"
    value=""
  />
  <button
    className="ColorEdit"
    onClick={[Function]}
    tabIndex={0}
    type="button"
  >
    Blue
    <span
      className="ColorDisplay"
      style={
        Object {
          "backgroundColor": "#289de9",
        }
      }
    />
    <span
      onKeyUp={[Function]}
      role="button"
      tabIndex={0}
      title="More"
    >
      <img
        alt="More"
        className="Arrow"
        src="/svgs/v.svg"
        style={
          Object {
            "height": "1em",
            "width": "1em",
          }
        }
      />
    </span>
  </button>
</li>
`;
